File: jiyeon5/week01/BOJ_2805.py
'''
푼 시간 : 27분
성능 : 메모리 149780KB, 시간 4824ms
주의할 점 :
    input으로 입력받고 숫자는 int형으로 변환 해주기
'''


# 입력 데이터 개수로 인한 시간초과 방지하기 위해 라이브러리 추가
import sys

# N: 나무의 수, M: 상근이가 집으로 가져가려고 하는 나무의 길이 입력 받기
N, M = map(int,input().split())
# 나무의 높이 입력 받기
trees = list(map(int,sys.stdin.readline().split()))

start = 0
end = max(trees) # 나무들의 높이 중 최대값을 초기 끝점으로 세팅

# 절단하는 높이 (구해야하는 값)
cut = 0
# 이분탐색
while start <= end:
    mid = (start + end) // 2

    sum = 0
    # 절단 후 가져가는 나무 높이 확인
    for t in trees:
        if t > mid: # 가져갈 수 있는 나무가 있다면 sum에 추가
           sum += t - mid

    # 절단 후 가져가는 높이가 M보다 크다면 절단 높이 저장 후 절단 높이 올리기
    if sum >= M:
        cut = mid
        start = mid +1
    # 절단 후 가져가는 높이가 M보다 작다면 절단 높이 낮추기
    else:
        end = mid -1

# 절단 높이 최대값 출력
print(cut)

File: teamzamong/week01/BOJ_2805.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
 * BOJ 2805: 나무 자르기
 * - 메모리: 123276KB
 * - 시간: 548ms
 *
 * [문제]
 * N개의 나무가 주어졌을 때, 절단기를 땅으로부터 H미터로 하여 나무를 자르려고 한다.
 * 이때, 최소 M 미터의 나무를 집에 가져가기 위해서 절단기에 설정할 수 있는 높이의 최댓값.
 *
 * [풀이]
 * - Parametric Search: 조건을 만족하는 최솟값/최댓값을 구하는 문제를 결정 문제로 변환해 이분탐색을 수행하는 방법
 * => 절단기의 높이가 H일 때, 가져갈 수 있는 나무의 길이가 M 이상인가 아닌가?
 *
 * 이분탐색의 범위를 인덱스가 아닌 H의 최댓값과 최솟값으로 설정해 풀이하는 문제
 *
 * [느낀 점]
 * - 강의에서 마지막에 다룬 문제랑 같은 문제였다... 강의를 듣고 풀자!!
 * - Array.sort()를 한 이후에 이분탐색을 시도했을 경우 1000ms 정도 나왔는데, 최댓값만 구해서 진행한 경우 500ms대로 줄어들었다.
 *   나무길이-절단기높이 < 0인 경우 때문에 정렬을 했던 건데, 정렬(O(NlogN))보다 O(N)으로 반복문 돌리는 게 더 효율적인 게 맞다 :(
 */

public class BOJ_2805 {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        long maxTree = 0;
        long[] trees = new long[N];
        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < N; ++i) {
            trees[i] = Long.parseLong(st.nextToken());
            if (trees[i] > maxTree) maxTree = trees[i];
        }

        System.out.println(binarySearch(trees, M, maxTree));
    }

    private static long binarySearch(long[] trees, int M, long max) {
        long start = 0;
        long end = max;

        while (start < end) {
            long mid = (start + end + 1) / 2; // +1 안 해주면 end-start = 1일 때 무한 루프 발생 가능

            if (parametricSearch(trees, mid, M)) {
                start = mid;
            } else {
                end = mid - 1; // mid 이상인 값은 답이 될 수 없음
            }
        }
        return start;
    }

    private static boolean parametricSearch(long[] trees, long cut, int M) {
        long total = 0;

        for (int i = trees.length - 1; i >= 0; --i) {
            if (trees[i] <= cut) continue;
            total += trees[i] - cut;
        }

        return total >= M;
    }

}
